drivers/starburst/src/metabase/driver/implementation/execute.clj

```diff
@@ -187,16 +191,50 @@
         (.setFetchDirection stmt ResultSet/FETCH_FORWARD)
         (catch Throwable e
           (log/debug e (trs "Error setting prepared statement fetch direction to FETCH_FORWARD"))))
-      (sql-jdbc.execute/set-parameters! driver stmt params)
-      (proxy [java.sql.PreparedStatement] []
-        (executeQuery []
-          (try
-            (let [rs (.executeQuery stmt)]
-              (remove-impersonation conn)
-              rs)
-              (catch Throwable e (handle-execution-error e))))
-        (setMaxRows [nb] (.setMaxRows stmt nb))
-        (close [] (.close stmt)))
+      ; Metabase is sometimes calling getParametersMetaData() on the prepared statement in order to count
+      ; the number of parameters and verify they are correct with what is expected
+      ; This unfortunately defeats the purpose of using EXECUTE IMMEDIATE as it forces the JDBC driver
+      ; to call an explicit PREPARE. However this call is optional, so the solution is to create a proxy
+      ; which defines all methods used by Metabase *except* for metadata methods. When Metabase tries
+      ; to count the number of parameters:
+      ; - The proxy issues an exception as getParametersMetaData() is not defined
+      ; - Metabase catches the exception and does not perform the check
+      ; - An invalid query is sent to Trino, which fails with a "Incorrect number of parameters" message
+      ; - This message is caught by the driver and replaced with the exact same Metabase message
+      ; In the end, the exact same message is presented to the user when the number of arguments is
+      ; incorrect except we now execute the query to display the error message
+      (let [ps (proxy [java.sql.PreparedStatement] []
```

Comment:
Should we be using this conditionally, only when optimized prepared statements are enabled?

Reply:
I updated the PR to use the old code when `TrinoConnection.useExplicitPrepare()` returns true and the new code otherwise.